NewXKitBot 2.0

Wolvan opened the pull request:
I decided to do a full rewrite of the bot, in turn cleaning it up and making it easier to use and to configure

Wolvan committed:
Clear out everything

Since this is a rewrite from scratch, clear out everything and restore a
clean workspace

Wolvan committed:
Create EventEmitter-based IRC- and Gitter-Clients

These 2 clients are wrappers for the bigger node-irc and node-gitter
modules that do most of the heavy lifting. Since only a few events and
functions are needed from each, these 2 wrapper modules gives convenient
access to them and make opening a connection easy.

Wolvan committed:
Create bot-core main.js

This is the bot's core to bootstrap it. Config loading and, for now,
GitterIRCBridge-initializing are done from this core file

Wolvan committed:
Do "npm init" and create package.json

Run "npm init" to create the package.json for easy bot installation and
meta-data

bvtsang commented on README.md:
```diff
@@ -1,3 +1,24 @@
-[![Gitter](https://badges.gitter.im/Join%20Chat.svg)](https://gitter.im/new-xkit/NewXKitBot?utm_source=badge&utm_medium=badge&utm_campaign=pr-badge&utm_content=badge)
\ No newline at end of file
+### What is New-XKit Bot?
+New-XKit Bot is the team behind [New XKit](https://github.com/new-xkit/XKit)'s helping little assistant, written in JavaScript for [Node.js](https://nodejs.org/). It's here to make the lives of the New-XKit team just a tiny bit easier!
```
This could be misread as "(New-XKit Bot is the team behind New XKit)'s helping little assistant" rather than "New-XKit Bot is (the team behind New XKit)'s helping little assistant". For clarity, I would rewrite this as

> New-XKit Bot is New XKit team's little assistant

bvtsang commented on README.md:
```diff
@@ -1,3 +1,24 @@
+### How do I contribute?
+1. Fork the repository
+2. Make your changes in the repository
+3. Open a PR with your changes
```
For readers who don't know what a PR is: it's a pull request. I know we're hosting this on Github, but I don't think all Github visitors know what a PR is.

Wolvan commented:
Made your changes top the README.md, @bvtsang

0xazure commented on README.md:
```diff
@@ -1,3 +1,24 @@
+### How do I run it myself?
+1. Install [Node.js](https://nodejs.org/)
+2. Clone this repository into a folder
+3. Copy the ``default.js`` file in the ``config`` subfolder and rename it to ``local.js``
```
GitHub's Markdown Basics document recommends [using single backticks (`)](https://help.github.com/articles/markdown-basics/#code-formatting).

0xazure commented on config/default.js:
```diff
@@ -0,0 +1,168 @@
+	GitterIRCBridges: [
+		{
+			gitter: {
+				token: process.env.NXB_GITTER_TOKEN,
```
No fallback or default (even if fake) if `NXB_*` vars are not defined in `process.env`.  I would recommend investigating [dotenv](https://www.npmjs.com/package/dotenv).

nightpool replied:
should refuse to start if you're missing a gitter token. other config options should generate defaults (which would then be saved)

Wolvan replied:
@0xazure it does indeed fail to start if these are not given. It errors out with either `Gitter Token or Room not specified!` or `Missing argument! You need to specify host, port, name, channel and welcome message!`

0xazure commented on modules/IRCChat.js:
```diff
@@ -0,0 +1,138 @@
+
+var IRCConnections = {};
+
+function IRCChat(host, server_port, name, pass, channel, WelcomeMsg, isSecure, cmd_prefix) {
```
When you get to this many params, you're probably better off using a config object instead of worrying about function parameter order.

0xazure commented on modules/GitterIRCBridge.js:
```diff
@@ -0,0 +1,106 @@
+			if (command_data.not_elevated) {
+				command_data.callback(from, args, irc, gitter);
+			} else {
+				if(irc.getUserRank(from) === "@") {
```
This is a poor method of authenticating privileged commands.  Unless the IRC channel permission levels are our source of truth (which I don't think they should be), we should at least be [checking against GitHub's Teams](https://developer.github.com/v3/orgs/teams/#get-team-membership) to ensure the user has the correct permissions.

Oddly, Gitter doesn't seem to provide any endpoints for determining who has what level of permissions in a given room.

nightpool replied:
maybe just wib hardcode users for now? we should be checking against github teams for sure, but if that's too difficult I'm fine on punting.

0xazure replied:
I'd be fine with pulling in a list of users set as an env var on boot in the interim, rather than hard-code names into a public project.

nightpool replied:
Okay? I don't really have a preference either way but... its pretty trivial to figure out who's involved with the new xkit project haha

Wolvan replied:
@0xazure I don't really see the problem here? Our team is OP in the IRC Channel (which will always be the rank `@`). This just verifies that whoever runs an elevated command has OP permissions

0xazure commented on LICENSE.md:
```diff
@@ -0,0 +1,344 @@
+
+---------------------------------------------------------------------------
+
+###            How to Apply These Terms to Your New Programs            ###
```
You need to actually apply these terms to the project.  This still has the example companies/details in it.

nightpool replied:
yeah remove everything after this line, for a start.

0xazure commented:
The lack of in-line documentation makes things a little tricky to understand what's going on, particularly in the cases where RegEx is being [used to transform](https://github.com/new-xkit/NewXKitBot/pull/4/files#diff-393abcf6328293f30752163c58f362f6R30) the messages.

I would also still like to see thorough documentation as to how the NewXKit bot is set up and run on our infrastructure.  It doesn't necessarily need to be a part of this project, but formal deployment instructions would be helpful.

Generally, having monolithic do-everything files make it more difficult to reason about a project.  Splitting the code up by concerns and making things more modular would go a long way to help readability and understanding.

These have just been my initial thoughts and comments, and do not constitute a formal code review.

nightpool commented on README.md:
```diff
@@ -1,3 +1,24 @@
-# NewXKitBot
+# New-XKit Bot
```
I think we standardized on New XKit in prose and new-xkit in code, fyi.

nightpool commented on config/default.js:
```diff
@@ -0,0 +1,168 @@
+					
+					list: {
+						not_elevated: true,
+						callback: function(user, args, gitter, irc) {
```
super-nitpick: callback doesn't describe the interface well. Maybe "run"?

nightpool commented on config/default.js:
```diff
@@ -0,0 +1,168 @@
+							if (users.length <= 1) {
+								msg = "[GitterBridge] There are no users online.";
+							} else {
+								for (var i = 0; i < users.length; i++) {
```
prefer map over for.

nightpool commented on config/default.js:
````diff
@@ -0,0 +1,168 @@
+						not_elevated: true,
+						callback: function(user, args, gitter, irc) {
+							irc.whois(args[0], function(username, whoisinfo) {
+								gitter.say("[GitterBridge] Whois for " + username + ":\n```\n" + JSON.stringify(whoisinfo, null, "\t") + "\n```");
````
prefer 120 line length

0xazure replied:
Is deep nesting even necessary here?  Why not define the `commands` object separately and do an assignment?

nightpool replied:
True

On Thu, Nov 19, 2015 at 11:44 PM 0xazure notifications@github.com wrote:

> In config/default.js
> https://github.com/new-xkit/NewXKitBot/pull/4#discussion_r45435541:
> 
> > -                           } else {
> > -                               for (var i = 0; i < users.length; i++) {
> > -                                   if (users[i] === irc.getUsername()) {
> > -                                       users.splice(i, 1);
> > -                                   }
> > -                               }
> > -                               msg = "[GitterBridge] There are " + users.length + " User(s) online: " + users.join(", ");
> > -                           }
> > -                           gitter.say(msg);
> > -                       }
> > -                   },
> > -                   whois: {
> > -                       not_elevated: true,
> > -                       callback: function(user, args, gitter, irc) {
> > -                           irc.whois(args[0], function(username, whoisinfo) {
> > -                               gitter.say("[GitterBridge] Whois for " + username + ":\n`\n" + JSON.stringify(whoisinfo, null, "\t") + "\n`");
> 
> Is deep nesting even necessary here? Why not define the commands object
> separately and do an assignment?
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/new-xkit/NewXKitBot/pull/4/files#r45435541.

nightpool commented on config/default.js:
```diff
@@ -0,0 +1,168 @@
+					xcloud: {
+						not_elevated: true,
+						callback: function(user, args, gitter, irc) {
+							var msg = "XCloud Migration Guide: http://portal.new-xkit.com/assets/migration/xcloud-migration.html";
```
should be a config, tbh.

nightpool replied:
okay this technically is a "config" but ideally this would be a property of the object I guess, not a local variable.

Wolvan replied:
Can you elaborate? I don't quite know what you mean

nightpool commented on modules/GitterIRCBridge.js:
```diff
@@ -0,0 +1,106 @@
+		gitter.say("`" + from + "` " + msg);
+	});
+	
+	irc.on("nick_changed", function(old_nick, new_nick) {
```
should probably be separate objects like the commands are, but fine to WIB/punt for now

Wolvan replied:
What should be seperate objects?

nightpool replied:
each event that we're handling from a gitter/irc client

Wolvan replied:
Not sure why, since this module is meant to bridge the 2 services having pre-defined bridging events seems logical to me

nightpool commented on modules/GitterIRCBridge.js:
```diff
@@ -0,0 +1,106 @@
+	);
+	
+	gitter.on("message", function(from, msg) {
+		msg = msg.replace(new RegExp(bridge_config.github.org + "/" + bridge_config.github.repo + "#", "gi"), "https://github.com/" + bridge_config.github.org + "/" + bridge_config.github.repo + "/issues/");
```
this should probably be separated into a few different functions where it makes sense.

nightpool commented on modules/GitterIRCBridge.js:
```diff
@@ -0,0 +1,106 @@
+			irc.say(from, msg);
+		}
+	}).on("message_update", function(from, msg) {
+		msg = msg.replace(new RegExp(bridge_config.github.org + "/" + bridge_config.github.repo + "#", "gi"), "https://github.com/" + bridge_config.github.org + "/" + bridge_config.github.repo + "/issues/");
```
code duplication!! should have one function onMessage(from, msg) that gets passed instead of these two anonymous functions, no?

nightpool commented on config/default.js:
```diff
@@ -0,0 +1,168 @@
+				nick: process.env.NXB_IRC_NICK,
+				password: "None",
+				channel: process.env.NXB_IRC_CHANNEL,
+				welcome_message: "Welcome to the Live Support for New XKit. If you did not yet set a username, please use /nick <Username> to do so.\nJust go ahead and ask whatever questions you have! If we do not respond immediately, please be patient. We might be away/busy/asleep. Once we have time, we'll get back to you.",
```
line length (it really does make your code more readable!!

nightpool commented on modules/IRCChat.js:
```diff
@@ -0,0 +1,138 @@
+		self.emit("kill", nick, reason, channels);
+	});
+	
+	self.say = function(from, message) {
```
`self.say`? shouldn't these just be normal functions on the object?

0xazure replied:
`self` here is `this`, and is then [mixed in with `events.EventEmitter`](https://github.com/new-xkit/NewXKitBot/pull/4/files#diff-ab2f83a50ad2e1c1b26d9ce0a36350c7R136).  I'm not particularly fond of this architecting though, because it's too confusing.  ~~Not to mention we're mixing `new` (when we [instantiate instances of IRCChat](https://github.com/new-xkit/NewXKitBot/pull/4/files#diff-393abcf6328293f30752163c58f362f6R17)) and `util.inherits`.~~ Internal implementation detail.

nightpool replied:
@0xazure is there any good reason these methods need to be dynamically constructed? rather then defined in the object?

0xazure replied:
I'm not even entirely sure what's going on here, because unless I'm reading things wrong, `IRCChat` is a function apparently with sub-functions, and doesn't return anything?  So your guess, @nightpool, is as good as mine.

Wolvan replied:
[That's OOP construction in JS for you](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Introduction_to_Object-Oriented_JavaScript#Custom_objects). If you assign something to `this` in a JS function and then use `new whatever()`, it automatically returns an instance of the class with all the `this.` attached stuff as methods/properties.

I could also do `IRCChat.prototype.function_name` outside of the function itself, but I prefer having the code for the entire object inside

0xazure replied:
Just because JS allows for it doesn't make it good practice.  I shouldn't have to think about unintuitive JS internals when I'm reading a function/object/whatever definition.  Making things explicit, obvious, and clear is better than being "clever".  Code shouldn't be "clever", it should be boring.

Wolvan replied:
What do I have to do instead? Completely scrap any kind of OOP? I didn't try to make the code clever, I tried to make it easy to use

nightpool replied:
@Wolvan the code should just be IRCClient = {say: function(){}, notice: function(){}, etc...} like any other javascript object.

0xazure replied:
Well, using explicit `return` statements would be a good start.  Choosing [prototypal inheritance **or** classical inheritance](http://stackoverflow.com/a/19640910) instead of mixing them would also help clarity.

Wolvan replied:
That would not allow instantiating multiple IRC connections though, at least not as easily as right now (`new IRCChat()`)

nightpool commented on package.json:
```diff
@@ -1,26 +1,33 @@
+    "bot",
+    "xkit",
+    "new",
+    "xkit",
```
xkit twice

nightpool commented on .gitignore:
```diff
@@ -1,7 +1,31 @@
-# Bot Config: Contains highly sensitive data, NEVER Commit
-newxkitbot.conf.js
+# Directory for instrumented libs generated by jscoverage/JSCover
+lib-cov
```
do we have lib-cov/coverage?

0xazure replied:
Looks like this is a (partial?) copy-paste of [GitHub's Node.gitignore](https://github.com/github/gitignore/blob/da00310ccba9de9a988cc973ef5238ad2c1460e9/Node.gitignore)

Wolvan replied:
It is GitHub's auto-generated .gitignore, yes. I just added the config/local and storage subfolder

nightpool commented:
would be nice to have this linted

0xazure commented:
> would be nice to have this linted

I'd probably recommend [XO](https://github.com/sindresorhus/xo) for a Just Works :tm: linter, if we don't want to fuss with config options, otherwise [ESLint](https://github.com/eslint/eslint).

Wolvan committed:
Re-init repository

After clearing out the old code, re-init the repository with a proper
README.md, a LICENSE.md (GPL-2.0) and a .gitignore for NodeJS projects

Wolvan committed:
Apply license properly

Put the license information in each source file the way GPL-2.0 says
and remove the "How to apply this license" part from LICENSE.md

Wolvan committed:
Create Gitter-IRC-Bridge module

This module combines the GitterChat and the IRCChat module to create a
functional bridge between these 2 services. Each GitterIRCBridge
represents a single pair of IRC and Gitter Channel.

Wolvan committed:
Create a default config

Should be self-explanatory, write a default configuration file
which can be extended through a local.js.

nightpool commented on modules/GitterIRCBridge.js:
```diff
@@ -0,0 +1,125 @@
+					} else {
+						command.not_allowed(user, gitter, irc);
+					}
+				}.bind(this, from, args, command_data));
```
unnecessary bind.

Wolvan replied:
Actually, that bind is necessary iirc. Since I am doing an async call here and I have 2 async calls collide it can lead to issues. I had that before

nightpool replied:
as long as you don't use this in your anonymous functions, you don't need to use bind. Also, remove the (user, args, command) parameters from the argument list, they're in the closure context of the function

Wolvan replied:
I have from, and args and command_data I use though. The loop progresses, overwriting those variables, so when a callback gets called it messes up. See Screenshot
![image](https://cloud.githubusercontent.com/assets/1437010/11306788/366e6156-8fb4-11e5-9845-0826d8740ea7.png)
This happened to me before doing the bind

nightpool replied:
right, but from, args and command_data don't change—they're static to the anonymous function context that's generated on line 96

0xazure commented on config/default.js:
```diff
@@ -0,0 +1,189 @@
+					not_elevated: true,
+					run: function(user, args, gitter, irc) {
+						gitter.say(
+							"#Commands\n" +
```
If the commands are already defined as objects, why aren't we auto-generating the list of commands?  Would be simple to add a `description` prop to each one instead of hard-coding them here.

Wolvan replied:
You know what? I absolutely didn't think of that, good idea.
Probably gonna do that in another PR after this one though.

0xazure commented on modules/GitterChat.js:
```diff
@@ -0,0 +1,78 @@
+				if (message.operation === "create") {
+					self.emit("message", message.model.fromUser.username, message.model.text, message);
+				} else if (message.operation === "update") {
+					self.emit("message_update", message.model.fromUser.username, message.model.text, message);
```
Code duplication with L52.

Wolvan replied:
It's 2 different events that get fired here, just the same arguments

0xazure replied:
I understand that they're different events.  My point about duplication stands.

A refactoring:

``` javascript
var event
if (message.operation === "create") {
  event = "message"
} else if (message.operation === "update") {
  event = "message_update"
}
self.emit(event, message.model.fromUser.username, message.model.text, message);
```

nightpool replied:
right that's why its called code duplication, because you're duplicating the code between the two lines in a place where it makes sense to be using a better abstraction

Wolvan replied:
Fair point

0xazure commented on modules/GitterChat.js:
```diff
@@ -0,0 +1,78 @@
+				var command = cmd_array.splice(0,1)[0].toLowerCase();
+				
+				self.emit("command", command, message.model.fromUser.username, cmd_array, message);
+				self.emit("command#" + command, message.model.fromUser.username, cmd_array, message);
```
Why are these two lines nearly identical?  What's the difference between `command` and `command#`?

Wolvan replied:
`command` fires on every command, the # is kind of like a filter if you want. Example: I have 3 commands, `!foo`, `!bar` and `!baz`.
If I do `.on("command")`, it fires everytime I do !foo, !bar or !baz and I would have to do switch statements to determine the command's function.
If I hook for example `.on("command#baz")`, the event only gets called when I run !baz

0xazure replied:
Hooking `.on("command#baz")` doesn't register as a general `command` event, in addition to a `!baz` command?

Wolvan replied:
Nope.
